src/test/unit/test_main_utils.py

```diff
@@ -62,3 +75,107 @@ def test_create_request_body_take_error_first(dummy_span, error_span, function_e
     assert _create_request_body(input, True, size) == json.dumps(
         [function_end_span, error_span, dummy_span, dummy_span]
     )
+
+
+@pytest.mark.parametrize(
+    ("var", "expected"),
+    [
+        ("aaa", "aaa"),  # Short.
+        ("a" * (Frame.MAX_VAR_LEN + 1), "a" * Frame.MAX_VAR_LEN + "..."),  # Long.
+        (["a", "b"], str(["a", "b"])),  # Type not str.
+    ],
+)
+def test_frame_truncate_var(var, expected):
+    assert Frame._truncate_var(var) == expected
+
+
+@pytest.mark.parametrize(
+    ("f_locals", "expected"),
+    [
+        ({}, {}),  # Empty.
+        ({"a": "b"}, {"a": "b"}),  # Short.
+        ({"a": "b" * (Frame.MAX_VAR_LEN + 1)}, {"a": "b" * Frame.MAX_VAR_LEN + "..."}),  # Long.
+        # Some short, some long.
+        (
+            {
+                "l1": "l" * (Frame.MAX_VAR_LEN + 1),
+                "s1": "s",
+                "l2": "l" * (Frame.MAX_VAR_LEN + 1),
+                "s2": "s",
+            },
+            {
+                "l1": "l" * Frame.MAX_VAR_LEN + "...",
+                "s1": "s",
+                "l2": "l" * Frame.MAX_VAR_LEN + "...",
+                "s2": "s",
+            },
+        ),
+        ({"a": ["b", "c"]}, {"a": str(["b", "c"])}),  # Not str.
+        ({"f": list}, {}),  # Function.
+    ],
+)
+def test_frame_truncate_locals(f_locals, expected):
+    assert Frame._truncate_locals(f_locals) == expected
+
+
+def test_frame_truncate_locals_pass_max_vars_size():
+    f_locals = {i: "i" for i in range(MAX_VARS_SIZE * 2)}
+    actual = Frame._truncate_locals(f_locals)
+    assert len(actual) < MAX_VARS_SIZE
+
+
+def test_extract_frames_from_exception():
+    def func_a():
+        a = "A"  # noqa
+        func_b()
+
+    def func_b():
+        one = 1
+        zero = 0
+        one / zero
+
+    try:
+        e = "E"  # noqa
+        func_a()
+    except Exception:
+        frames = extract_frames_from_exception()
+
+    assert frames[0].function == "func_b"
+    assert frames[0].variables == {"one": "1", "zero": "0"}
+    assert frames[1].function == "func_a"
+    assert frames[1].variables == {"a": "A"}
+    assert frames[2].function == "test_extract_frames_from_exception"
+    assert frames[2].variables == {"e": "E"}
+
+
+def test_extract_frames_from_exception_pass_max_vars_size():
```

Comment:
another test that you can add is a single parameter which is huge
`d = {i: "1" for i in range(MAX_VARS_SIZE)}`